xdsl/dialects/accfg.py

```diff
@@ -0,0 +1,310 @@
+            )
+        # that the token and my accelerator match
+        assert isinstance(self.token.type, TokenType)
+        if self.token.type.accelerator != self.accelerator:
+            raise VerifyException(
+                "The token's accelerator does not match the launch accelerator!"
+            )
+
+        # that the token is used
+        if len(self.token.uses) != 1 or not isinstance(
+            next(iter(self.token.uses)).operation, AwaitOp
+        ):
+            raise VerifyException("Launch token must be used by exactly one await op")
+        # TODO: allow use in control flow
+
+
+@irdl_op_definition
+class AwaitOp(IRDLOperation):
+    """
+    Wait until the launched operation finishes.
+    Awaits the token emitted by an accfg.launch operation.
+    """
+
+    name = "accfg.await"
+
+    token = operand_def(TokenType)
+
+    def __init__(self, token: SSAValue | Operation):
+        super().__init__(operands=[token])
+
+
+@irdl_op_definition
+class SetupOp(IRDLOperation):
+    """
+    accfg.setup writes values to a specific accelerators configuration and returns
+    a value representing the currently known state of that accelerator's config.
+
+    If accfg.setup is called without any parameters, the resulting state is the
+    "empty" state, that represents a state without known values.
+    """
+
+    name = "accfg.setup"
+
+    values = var_operand_def(Attribute)  # TODO: make more precise?
+    """
+    The actual values used to set up the CSRs
+    """
+
+    in_state = opt_operand_def(StateType)
+    """
+    The state produced by a previous accfg.setup
+    """
+
+    out_state = result_def(StateType)
+    """
+    The CSR state after the setup op modified it.
+    """
+
+    param_names = prop_def(ArrayAttr[StringAttr])
+    """
+    Maps the SSA values in `values` to accelerator parameter names
+    """
+
+    accelerator = prop_def(StringAttr)
+    """
+    Name of the accelerator this setup is for
+    """
+
+    irdl_options = [AttrSizedOperandSegments(as_property=True)]
+
+    def __init__(
+        self,
+        vals: Sequence[SSAValue | Operation],
+        param_names: Iterable[str] | Iterable[StringAttr],
+        accelerator: str | StringAttr,
+        in_state: SSAValue | Operation | None = None,
+    ):
+        if not isinstance(accelerator, StringAttr):
+            accelerator = StringAttr(accelerator)
+
+        param_names_tuple: tuple[StringAttr, ...] = tuple(
+            StringAttr(name) if isinstance(name, str) else name for name in param_names
+        )
+
+        super().__init__(
+            operands=[vals, in_state],
+            properties={
+                "param_names": ArrayAttr(param_names_tuple),
+                "accelerator": accelerator,
+            },
+            result_types=[StateType(accelerator)],
+        )
+
+    def iter_params(self) -> Iterable[tuple[str, SSAValue]]:
+        return zip((p.data for p in self.param_names), self.values)
```

Comment:
If we expect the values to have the same length always, then it seems like a good idea to add `strict=True` here.